### PR TITLE
Vulnerabilities refactor continued

### PR DESCRIPTION
**Pull request by nightfurys**
This PR re-introduces deduping vulnerabilities that was skipped in the refactor PR #1000 One main difference with this addition is the where the dedup is performed. It was originally in external API and slightly differing variants sprinkled through out the system (policy eval, other services). This update fixes the issue by moving the dedup to get-image-vulnerabilities path - this guarantees every client gets the same information without the dedup know-how. Added some unit tests as well

### CODE REVIEW

File: anchore_engine/services/policy_engine/engine/vulns/providers.py [hunk 1]
```diff
@@ -59,29 +54,40 @@ class VulnerabilitiesProvider:
...
+    @abstractmethod
     def load_image(self, **kwargs):
         """
         Ingress the image and compute the vulnerability matches. To be used in the load image path to prime the matches
         """
-        raise NotImplementedError()
+        ...
+
+    @abstractmethod
+    def get_image_vulnerabilities_json(self, **kwargs) -> json:
```
Comment (zhill) [inline anchore_engine/services/policy_engine/engine/vulns/providers.py:65]: I'm not sure 'json' is what you intended here. I think this is using the json module as the type.

File: anchore_engine/services/policy_engine/engine/vulns/providers.py [hunk 2]
```diff
@@ -100,14 +106,26 @@ def load_image(self, image: Image, db_session, cache=False):
...
+        self,
+        image: Image,
+        db_session,
+        vendor_only: bool = True,
+        force_refresh: bool = False,
+        cache: bool = True,
+    ) -> json:
+        return self.get_image_vulnerabilities(
+            image, db_session, vendor_only, force_refresh, cache
+        ).to_json()
```
Comment (zhill) [inline anchore_engine/services/policy_engine/engine/vulns/providers.py:119]: This seems outside the scope of the provider: formatting the output object.

**Review by zhill — COMMENTED** [review]
Overall the logic makes sense, I just have comments on how to simplify things a bit. The dedup logic may be simplified by using object hashing and identity to support set operations, but that is a nit, not blocking.

File: anchore_engine/services/policy_engine/engine/vulns/providers.py (hunk 1 above)
Comment (nightfurys) [reply anchore_engine/services/policy_engine/engine/vulns/providers.py:65]: That's incorrect, not sure how it made it there

File: anchore_engine/services/policy_engine/engine/vulns/providers.py (hunk 2 above)
Comment (nightfurys) [reply anchore_engine/services/policy_engine/engine/vulns/providers.py:119]: I had the jsonification outside of the provider scope initially but introduced it as an optimization for the grype-integration. When the report is loaded from cache, get_image_vulnerabilities_json() saves the additional steps to deserialize the json into object model and serialize it back again into json again. get_image_vulnerabilities() method will return the object model response, currently it's used by the policy evaluation gate to load vulnerabilities into the context

File: anchore_engine/services/policy_engine/engine/vulns/dedup.py [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+            "vuln_id",
+            "vuln_namespace",
+            "pkg_name",
+            "pkg_version",
+            "pkg_type",
+            "pkg_path",
+            "rank",
+        ],
+    )
+    VulnerabilityRankMatchTuple = namedtuple(
```
[the remark this answers is not in the document]
Comment (nightfurys) [reply anchore_engine/services/policy_engine/engine/vulns/dedup.py:48]: Tuples play well with hashed collections like sets and make it easier for comparison. Dataclasses might be able to deliver the same functionality but might need overriding hash and eq functions
